Sds trustworthy jwt doc

pitlv2109 opened the pull request:
Changes in this PR:
* Update installation doc for SDS since trustworthy jwt is enabled by default.
* Remove the setup in the task and refer to the installation page to prevent future divergence.   

#4777

Do not merge until https://github.com/istio/istio/pull/15568 is merged. 

[ ] Configuration Infrastructure
[X] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

pitlv2109 commented:
linter error 
```
./public/about/notes/1.1.9/index.html
  *  External link https://github.com/istio/proxy/issues/2160 failed: 404 No error
```

Probably need to add some exception to the http proofer @geeknoid

quanjielin commented on content/docs/tasks/security/auth-sds/index.md:
```diff
@@ -49,14 +49,7 @@ This approach has the following benefits:
 
 * Set up Istio by following the instructions using
   [Helm](/docs/setup/kubernetes/install/helm/) with SDS setup and global mutual
-  TLS enabled:
```
why remove this ?

pitlv2109 replied:
We're already pointing the users to the installation page, I don't think we need to list out the commands here as well. This is to prevent out of date information on the sds task page when the installation page gets updated.

pitlv2109 replied:
Not sure if I understand your comment, the installation has that information as well, if you're talking about this command.
`helm template install/kubernetes/helm/istio --name istio --namespace istio-system \
    --values install/kubernetes/helm/istio/values-istio-sds-auth.yaml | kubectl apply -f -`

The `useNormalJwt` and `useTrustworthyJwt` part are just a caveat section if the users do not want to enable trustworthy jwts. They still need to run the helm command.

quanjielin replied:
I see, thanks.

pitlv2109 commented:
https://github.com/istio/istio/pull/15568 has been merged. This PR is ready now.

myidpt commented on content/docs/setup/kubernetes/install/helm/index.md:
```diff
@@ -41,6 +41,11 @@ and then further customize the configuration for your specific needs.
 
+1. For Istio with SDS enabled:
+  * Please use Kubernetes v1.12 or above.
+  * For non-GKE environments, please add [extra configurations](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/?origin_team=T382U9E4U#service-account-token-volume-projection) to your Kubernetes. You may
```
Does this mean I have the restart the apiserver?

pitlv2109 replied:
For upgrading from an existing cluster, I think so. But this is for the installation Istio page, so I assume no Istio control plane and Istio-controlled workloads have been deployed, so it should not matter. For an existing cluster, they will need to upgrade apiserver and Citadel agent (to use trustworthy jwt, since the default used to be legacy jwt).

istio-testing commented:
@pitlv2109: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
@pitlv2109: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio.io-presubmit | 6267d2b700f49a01cafa82aa2b144841b7036051 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/4701/istio.io-presubmit/286) | `/test istio.io-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@pitlv2109: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio.io-presubmit | 839304ee46788dee10cf159774d06884d168c9e8 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/4701/istio.io-presubmit/287) | `/test istio.io-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@pitlv2109: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio.io-presubmit | 9a50b8c95c9359af7399b4736908972b1294520e | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/4701/istio.io-presubmit/288) | `/test istio.io-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pitlv2109 commented:
@rcaballeromx PTAL. I added an announcement and some info about the sds change in the platform setup page.
@geeknoid a bunch of envoy links are 404-ing.

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio.io-presubmit | 3f545895ffa178d4cb76fd863b1e3448bc520872 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/4701/istio.io-presubmit/384) | `/test istio.io-presubmit`
istio.io-lint | 3f545895ffa178d4cb76fd863b1e3448bc520872 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/4701/istio.io-lint/18) | `/test istio.io-lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pitlv2109 commented:
@myidpt  @rlenglet  @rcaballeromx I updated the doc to reflect https://github.com/istio/istio/pull/16666. PTAL.

rcaballeromx commented on content/en/blog/2019/trustworthy-jwt-sds/index.md:
```diff
@@ -0,0 +1,37 @@
+In Istio 1.3, we have made a change to take advantage of improvements in Kubernetes in order to more
+securely issue certificates for workloads.
+
+When a Citadel Agent (the agent that runs on a node and is responsible for obtaining certificates)
```
@pitlv2109 I'm fine with us doing a separate security-specific glossary PR.

pitlv2109 replied:
Sounds good.

rcaballeromx commented on content/en/blog/2019/trustworthy-jwt-sds/index.md:
```diff
@@ -0,0 +1,37 @@
+securely issue certificates for workloads.
+
+When a Citadel Agent (the agent that runs on a node and is responsible for obtaining certificates)
+sends a Certificate Signing Request to Citadel to get a certificate for a workload,
```
I couldn't find `Certificate Signing Request` in that medium post. I would suggest we don't even capitalize SDS. The convention is not to capitalize the words of the abbreviation. See https://en.wikipedia.org/wiki/Certificate_signing_request
Be wary of using capitalization for emphasis, that should be avoided because readers will be confused by it.

pitlv2109 replied:
Got it. Thanks!

rcaballeromx commented on content/en/docs/setup/platform-setup/_index.md:
```diff
@@ -9,4 +9,4 @@ keywords: [platform-setup]
 {{< tip >}}
 Istio {{< istio_version >}} has been tested with these Kubernetes releases: {{< supported_kubernetes_versions >}}.
-{{< /tip >}}
+{{</ tip >}}
```
The correct syntax, AFAIK, is:
```suggestion
{{< /tip >}}
```
Did your syntax render correctly?

pitlv2109 replied:
You're right. It has no problem rendering because there was nothing after this text. Even if you leave the ending tag, it's still okay. That's why I didn't see any issue. Will make the change. Thanks!

rcaballeromx commented on content/en/docs/setup/install/helm/index.md:
```diff
@@ -42,6 +42,14 @@ and then further customize the configuration for your specific needs.
 
 1. [Install a Helm client](https://github.com/helm/helm/blob/master/docs/install.md) with a version higher than 2.10.
 
+**Optional:** To enable SDS in Istio, upgrade your Kubernetes cluster to at least 1.13.
```
Are these steps of the installation? Are these considerations the reader needs to make before installing? If they are, they should not be at the end of the procedure. Please include this information on the `Before you begin` section.
Please make sure that there is no duplication or repetition of the information. Do not add platform-specific information to the general installation flow, add it to the platform-specific instructions.

pitlv2109 replied:
I'll move it to the before you install section.

JimmyCYJ commented:
@myidpt Do we want to update the caveats section for control plane SDS in this PR or in a separate PR?
https://preliminary.istio.io/docs/tasks/security/auth-sds/#caveats

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio.io-presubmit | 3f545895ffa178d4cb76fd863b1e3448bc520872 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/4701/istio.io-presubmit/384) | `/test istio.io-presubmit`
istio.io-lint | 54dce73bf0d4d7dd6a651042df4d6867683e1951 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/4701/istio.io-lint/74) | `/test istio.io-lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

JimmyCYJ commented:
> @myidpt Do we want to update the caveats section for control plane SDS in this PR or in a separate PR?
> https://preliminary.istio.io/docs/tasks/security/auth-sds/#caveats

I will update that section in a separate PR.

rcaballeromx commented on content/en/blog/2019/trustworthy-jwt-sds/index.md:
```diff
@@ -0,0 +1,31 @@
+1. The tokens don't have important fields to limit their scope of usage, such as `aud` or `exp`. See [Bound Service Tokens](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/auth/bound-service-account-tokens.md) for more info.
+1. The tokens are mounted onto all the pods without a way to opt-out. See [Service Account Token Volumes](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/svcacct-token-volume-source.md) for motivation.
+
+The `trustworthy` JWTs were introduced in Kubernetes 1.12 to solve the aforementioned issues.
```
```suggestion
Kubernetes 1.12 introduces `trustworthy` JWTs to solve these issues.
```

rcaballeromx commented on content/en/blog/2019/trustworthy-jwt-sds/index.md:
```diff
@@ -0,0 +1,31 @@
+1. The tokens are mounted onto all the pods without a way to opt-out. See [Service Account Token Volumes](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/svcacct-token-volume-source.md) for motivation.
+
+The `trustworthy` JWTs were introduced in Kubernetes 1.12 to solve the aforementioned issues.
+However, Kubernetes didn't support the `aud` to be anything other than the API server audience until [Kubernetes 1.13](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.13.md).
```
```suggestion
However, Kubernetes support for the `aud` field to be different from the API server audience became available until [Kubernetes 1.13](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.13.md).
```

rcaballeromx commented on content/en/blog/2019/trustworthy-jwt-sds/index.md:
```diff
@@ -0,0 +1,31 @@
+
+The `trustworthy` JWTs were introduced in Kubernetes 1.12 to solve the aforementioned issues.
+However, Kubernetes didn't support the `aud` to be anything other than the API server audience until [Kubernetes 1.13](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.13.md).
+To better secure the mesh, Istio 1.3 only supports `trustworthy` JWTs and sets the `aud` to a specific value (for example, the value is `istio-ca` in Citadel) when you enable SDS.
```
```suggestion
To better secure the mesh, Istio 1.3 only supports `trustworthy` JWTs and requires a specific value for the `aud` field. 
For example, you must set the value to `istio-ca` in Citadel to enable SDS.
```

rcaballeromx commented on content/en/blog/2019/trustworthy-jwt-sds/index.md:
```diff
@@ -0,0 +1,31 @@
+To better secure the mesh, Istio 1.3 only supports `trustworthy` JWTs and sets the `aud` to a specific value (for example, the value is `istio-ca` in Citadel) when you enable SDS.
+Before upgrading your Istio deployment to 1.3 with SDS enabled, verify that you use Kubernetes 1.13 or later.
+
+Make the following considerations based on your platform of choice:
```
Replace the content from line 26-31 with the following suggestion:
```suggestion
See the [platform-specific setup](/docs/setup/platform-setup/) for any considerations specific to your platform.
```

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio.io-presubmit | 3f545895ffa178d4cb76fd863b1e3448bc520872 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/4701/istio.io-presubmit/384) | `/test istio.io-presubmit`
istio.io-lint | d6e310a93262cf3091f7abdafd02f872c0e3f6e6 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/4701/istio.io-lint/94) | `/test istio.io-lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pitlv2109 commented:
@rcaballeromx  I updated the PR to address some of your comments. For other comments, I'd like to keep the content as it is. For instance, I still want to mention about k8s version and vendor support tin this blog post. If the users go to the Azure page and find nothing about sds there, they will assume it will work out of the box, which may not be true. 

> Consider adding a call to action to the blog post too:
What do you want someone that reads the post to the end to do? Do you want readers to try SDS out? File the info for the future? Give you feedback? How?

The goal of this blog post is for the SDS change announcement like the title suggests. Doing anything is up to the users. This is not a persuasive post. 

I feel like we're just going back and forth with the wording and structure. At one point we were gonna do this, at another we switched to a different approach. There is no clear guideline anywhere. If you feel like this PR is far from being approved, feel free to draft another PR and I can review the technical details. This PR has taken a big chunk of my time (and your time to review it as well) :)

cc @myidpt @rlenglet for opinions.

rlenglet commented:
@rcaballeromx could you please clarify which of your comments are blockers, and which are just suggestions?

pitlv2109 commented:
@rcaballeromx Sorry if my previous comment sounds harsh, but it was simply an effort to save time for everyone. This goes without saying that I appreciate your time and feedback :)

myidpt commented on content/en/blog/2019/trustworthy-jwt-sds/index.md:
```diff
@@ -0,0 +1,31 @@
+1. The tokens are mounted onto all the pods without a way to opt-out. See [Service Account Token Volumes](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/svcacct-token-volume-source.md) for motivation.
+
+Kubernetes 1.12 introduces `trustworthy` JWTs to solve these issues.
+However, Kubernetes support for the `aud` field to be different than the API server audience became available until [Kubernetes 1.13](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.13.md).
```
```suggestion
However, Kubernetes' support for the `aud` field to be different than the API server audience became available until [Kubernetes 1.13](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.13.md).
```

rcaballeromx replied:
There's no need to have Kubenetes twice in the sentence also fixing minor verb tense issue. 
```suggestion
However, support for the `aud` field to have a different value than the API server audience didn't become available with [Kubernetes 1.13](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.13.md).
```

pitlv2109 replied:
@rcaballeromx The language is messing with me here. So the information that I'm trying to communicate here is k8s didn't support different audience until version 1.3. Is what you're suggesting saying that? I think I missed the `didn't` part in my current PR.

rcaballeromx replied:
Sorry wrong preposition crept in. Is this better:
```suggestion
However, support for the `aud` field to have a different value than the API server audience didn't become available until [Kubernetes 1.13](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.13.md).
```

myidpt commented on content/en/blog/2019/trustworthy-jwt-sds/index.md:
```diff
@@ -0,0 +1,31 @@
+
+Kubernetes 1.12 introduces `trustworthy` JWTs to solve these issues.
+However, Kubernetes support for the `aud` field to be different than the API server audience became available until [Kubernetes 1.13](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.13.md).
+To better secure the mesh, Istio 1.3 only supports `trustworthy` JWTs and requires a specific value for the `aud` field when you enable SDS.
```
```suggestion
To better secure the mesh, Istio 1.3 only supports `trustworthy` JWTs and requires the specific value `istio-ca` for the `aud` field when you enable SDS.
```

rcaballeromx replied:
```suggestion
To better secure the mesh, Istio 1.3 only supports `trustworthy` JWTs and requires the value of the `aud` field to be `istio-ca` when you enable SDS.
```

myidpt commented on content/en/blog/2019/trustworthy-jwt-sds/index.md:
```diff
@@ -0,0 +1,31 @@
+Kubernetes 1.12 introduces `trustworthy` JWTs to solve these issues.
+However, Kubernetes support for the `aud` field to be different than the API server audience became available until [Kubernetes 1.13](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.13.md).
+To better secure the mesh, Istio 1.3 only supports `trustworthy` JWTs and requires a specific value for the `aud` field when you enable SDS.
+For example, if the Certificate Authority is Citadel, Istio will set the `aud` to be `istio-ca` by default.
```
```suggestion
```

rcaballeromx commented on content/en/docs/setup/platform-setup/minikube/index.md:
```diff
@@ -12,9 +12,14 @@ keywords: [platform-setup,kubernetes,minikube]
+
+- Administrative privileges are required to run minikube.
+
+- To enable SDS in Istio, add [extra configurations](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection) to your Kubernetes. You may
```
```suggestion
- To enable SDS for your mesh, you must add [extra configurations](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection) to your Kubernetes deployment.
```

rcaballeromx commented on content/en/docs/setup/platform-setup/minikube/index.md:
```diff
@@ -12,9 +12,14 @@ keywords: [platform-setup,kubernetes,minikube]
+- Administrative privileges are required to run minikube.
+
+- To enable SDS in Istio, add [extra configurations](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection) to your Kubernetes. You may
+also want to refer to the [api-server page](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/) for the most up-to-date flag names.
```
```suggestion
Refer to the [`api-server` reference docs](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/) for the most up-to-date flags.
```
